Additions to js_string and js_array

glennsl opened the pull request:
This more or less brings js_array and js_string to es2015 compliance.

It has the following breaking changes:
* `Array.slice_copy` has been renamed `Array.sliceCopy`
* `Array.slice_start` has been renamed `Array.sliceFrom`
* `Array.lastIndexOf` has been renamed `Array.lastIndexOfFrom`
* `Array.lastIndexOf_start` has been renamed `Array.lastIndexOf`
* `String.concat` has been renamed `String.concatMany`
* A new non-spliced function has taken the place of the old `String.concat`

glennsl commented on jscomp/others/js_array.ml:
```diff
@@ -17,61 +17,110 @@
+external fillFromTo : 'a -> int -> int -> 'this = "" [@@bs.send.pipe: 'a t as 'this]
+
+(** https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/push *)
+external pop : 'a array -> 'a Js.undefined = "" [@@bs.send]
```
I'm thinking these should have pipe form too, but was hesitant to make the breaking change.

glennsl commented on jscomp/others/js_array.ml:
```diff
@@ -17,61 +17,110 @@
+(** https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/push *)
+external pop : 'a array -> 'a Js.undefined = "" [@@bs.send]
+external push : 'a array  -> 'a -> int = "" [@@bs.send]
+external pushMany : 'a array  -> 'a array -> int = "push" [@@bs.send] [@@bs.splice]
```
I'm thinking these should have pipe form too, but was hesitant to make the breaking change,

bobzhang commented:
hi @glennsl , are you aware of `[@@ocaml.deprecated "xx"]` maybe you can keep the old definition and add new one, so we can remove the deprecated in the next version?

glennsl commented:
I was not, but that's cool! I've added back the renamed functions with deprecation warnings, but this still leaves `String.concat` as a breaking change, and `push`/`pop` would be breaking if made into pipe form too.

glennsl commented:
Just noticed the `lastIndexOf` and `lastIndexFrom` signatures were off, so it's still a breaking change.

bobzhang commented on jscomp/others/js_array.ml:
```diff
@@ -17,61 +17,116 @@
-external toLocaleString : unit -> string  = "" [@@bs.send.pipe: 'a t as 'this]
-external concat : 'this -> 'this  = "" [@@bs.send.pipe: 'a t as 'this]
-external append : 'a -> 'this = "concat" [@@bs.send.pipe: 'a t as 'this]
+external isArray : 'a -> Js.boolean = "Array.isArray" [@@bs.val]
```
This is an es6 feature, shall we document it clearly?

bobzhang commented on jscomp/others/js_array.ml:
```diff
@@ -17,61 +17,116 @@
+*)
+external copyWithin : int -> 'this = "" [@@bs.send.pipe: 'a t as 'this]
+external copyWithinFrom : int -> int -> 'this = "copyWithin" [@@bs.send.pipe: 'a t as 'this]
+external copyWithinFromTo : int -> int -> int -> 'this = "copyWithin" [@@bs.send.pipe: 'a t as 'this]
```
add label argument?

bobzhang commented on jscomp/others/js_string.ml:
```diff
@@ -17,12 +17,12 @@
  * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)
 (* Copyright (C) 2015-2016 Bloomberg Finance L.P.
- * 
+ *
```
we have two copy right headers here again..

bobzhang commented on jscomp/test/js_string_test.js:
```diff
@@ -29,7 +29,7 @@ function eq(loc, _z) {
 eq('File "js_string_test.ml", line 13, characters 5-12', /* tuple */[
       "012",
-      "0".concat("1", "2")
+      "0".splice("1", "2")
```
this seems fishy

glennsl replied:
Sorry, that's been fixed, just missed commiting the generated js. I'll add more tests to avoid issues like this.

bobzhang commented:
This looks good to me in general, but it would be nice to have some docs/tests for each function

glennsl commented:
Ok, I've now squashed some more bugs, tweaked the API, added yet more functions and tests for everything. Once you give the OK I'll squash the commits and add the sign-off.

bobzhang commented on jscomp/others/js_array.ml:
```diff
@@ -17,61 +17,122 @@
+
+external shift : 'a Js.undefined = "" [@@bs.send.pipe: 'a t as 'this]
+
+external sort : 'this = "" [@@bs.send.pipe: 'a t as 'this]
```
can you document that it is functional sort or imperative sort. so we have some names `sortInPlace` for imperative ones

bobzhang commented on jscomp/others/js_array.ml:
```diff
@@ -17,61 +17,122 @@
 
-
 type 'a t = 'a array
+type 'a array_like
```
can you document where is `array_like` and `array_iter` used?

glennsl replied:
These were added primarily for the iterator functions, and if they're taken out there's currently no use for them. But there are many objects elsewhere that are array-like, `arguments`, `NodeList` and `HTMLCollection` to name a few. Without `array_like` and the `Array.from` function, these would all have to implement `Array.from` themselves as `toArray`. Unsure which is the right approach here.

bobzhang commented on jscomp/others/js_array.ml:
```diff
@@ -17,61 +17,122 @@
+external copyWithinFrom : to_:int -> from:int -> 'this = "copyWithin" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
+external copyWithinFromRange : to_:int -> start:int -> end_:int -> 'this = "copyWithin" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
+
+external fill : 'a -> 'this = "" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
```
I am thinking of to make it more explicit like 
`Js.Array.Es6.copyWithin` what do you think

glennsl replied:
I agree with making imperative functions explicit, but this specific case seems pretty explicit to me. I don't know what else it could mean. Unless we employ a more generic convention like `fillMut`, I don't know what I'd change this to.

bobzhang commented on jscomp/others/js_array.ml:
```diff
@@ -17,61 +17,122 @@
-external length : 'a array -> int = ""
-[@@bs.get]
-   
+external values : 'a array_iter = "" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
```
ah, so this is ES6 iterator, maybe we should not expose here, we need do some research to see how we could support iterator pattern in Buckle

glennsl commented on jscomp/others/js_array.ml:
```diff
@@ -17,61 +17,122 @@
+type 'a array_like
+type 'a array_iter = 'a array_like (* don't think this is very useful to implement wihtout language support *)
+
+external make : int -> unit Js.undefined array = "Array" [@@bs.new]
```
I included it for the ability to do things like `make 10 |> mapi (fun _ i -> i)`, but I agree `Array.make` is better.

bobzhang commented:
How about fillInPlace -- in gneral, in OCaml, people would expect an imperative function to return `unit`, here JS version is not returning unit (for chainning?), I think having a clear hint that it is imperative will be useful

reply@reply.github.com At: 02/14/17 17:22:15" data-digest="From: reply@reply.github.com At: 02/14/17 17:22:15" style="">
From: reply@reply.github.com At: 02/14/17 17:22:15
To: bucklescript@noreply.github.com
Cc: HONGBO ZHANG (BLOOMBERG/ 731 LEX), comment@noreply.github.com
Subject: Re: [bloomberg/bucklescript] Additions to js_string and js_array (#1203)


@glennsl commented on this pull request.   
In jscomp/others/js_array.ml: > +external unsafeFrom : 'a -> 'b array = "Array.from" [@@bs.val] (* es2015 *)
+external fromMap : 'a array_like -> ('a -> 'b [@bs]) -> 'b array = "Array.from" [@@bs.val] (* es2015 *)
+external unsafeFromMap : 'a -> ('b -> 'c [@bs]) -> 'c array = "Array.from" [@@bs.val] (* es2015 *)
+external isArray : 'a -> Js.boolean = "Array.isArray" [@@bs.val] (* es2015 *)
+(* Array.of: seems pointless unless you can bind *) (* es2015 *)
+
+external length : 'a array -> int = "" [@@bs.get]
+
+
+(* Mutator functions
+*)
+external copyWithin : to_:int -> 'this = "" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
+external copyWithinFrom : to_:int -> from:int -> 'this = "copyWithin" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
+external copyWithinFromRange : to_:int -> start:int -> end_:int -> 'this = "copyWithin" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
+
+external fill : 'a -> 'this = "" [@@bs.send.pipe: 'a t as 'this] (* es2015 *)
 
I agree with making imperative functions explicit, but this specific case seems pretty explicit to me. I don't know what else it could mean. Unless we employ a more generic convention like fillMut, I don't know what I'd change this to.  
—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or mute the thread.

glennsl commented:
Right, I'll do that then. I also tried to do a rebase, which only seems to have made things worse. Guess I have to `rebase -i` and force push. Will experiment.

glennsl commented:
Ok, rebase and force push seems to have gone well. And I've made come changes according to your comments.

bobzhang commented:
looks good, thanks! I will enable v6 so you can test latest API

glennsl commented:
I've never seen merlin show doc comments. But I've also never seen doc comments in the wild, so I don't know which shape they ought to have.
